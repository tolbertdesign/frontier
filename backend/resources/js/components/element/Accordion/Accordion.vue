<template>
  <div>
    <BCollapse
      :open="isOpen"
      class="card shadow-none border-t-2 border-grey-light"
    >
      <button
        slot="trigger"
        slot-scope="props"
        class="flex items-baseline w-full justify-between mb-px py-4 sm:py-4 px-4 md:px-4 cursor-pointer shadow-none focus:outline-none"
        @click="collapseClick"
      >
        <slot name="title">
          <p class="text-xl sm:text-2xl font-semibold">
            <span class="pl-1">{{ title }}</span>
          </p>
        </slot>
        <slot name="toggle">
          <span class="flex cursor-pointer items-center justify-center text-grey-dark">
            <FontAwesomeIcon :icon="props.open ? 'chevron-up' : 'chevron-down'" />
          </span>
        </slot>
      </button>
      <slot>
        <div class="p-2">
          <div class="modal-card-body bg-white p-4 md:p-6 text-sm sm:text-base md:text-lg">
            <p class="mb-4">
              {{ content }}
            </p>
          </div>
        </div>
      </slot>
    </BCollapse>
  </div>
</template>

<script>
import { FontAwesomeIcon } from '@fortawesome/vue-fontawesome'
export default {
  name: 'Accordion',
  status: 'prototype',
  release: '1.0.0',
  components: {
    FontAwesomeIcon,
  },
  props: {
    title: {
      type: String,
      default: '',
    },
    content: {
      type: String,
      default: '',
    },
    isOpen: {
      type: Boolean,
      default: false,
    },
    shouldReslick: {
      type: Boolean,
      default: false,
    },
  },
  methods: {
    collapseClick () {
      if (this.shouldReslick) {
        this.$nextTick(
          () => {
            if (navigator.userAgent.indexOf('MSIE') !== -1 || navigator.appVersion.indexOf('Trident/') > 0) {
              var evt = document.createEvent('UIEvents')
              evt.initUIEvent('reslicked', true, false, window, 0)
              window.dispatchEvent(evt)
            } else {
              window.dispatchEvent(new Event('reslicked'))
            }
          },
        )
      }
    },
  },
}
</script>
