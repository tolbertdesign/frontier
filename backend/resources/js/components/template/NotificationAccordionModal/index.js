import NotificationAccordionModal from './NotificationAccordionModal.vue'

export default NotificationAccordionModal
