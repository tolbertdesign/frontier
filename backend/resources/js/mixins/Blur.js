export default {
  methods: {
    blur () {
      document.getElementById('app').style.filter = 'blur(4px)'
    },
    unBlur () {
      document.getElementById('app').style.filter = 'none'
    },
  },
}
