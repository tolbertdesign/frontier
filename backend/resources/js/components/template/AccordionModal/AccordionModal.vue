<template>
  <div class="max-h-screen overflow-y-auto shadow-lg">
    <div class="modal-card rounded-xl bg-white shadow-lg">
      <header class="modal-card-head bg-white relative items-start">
        <slot name="header">
          <p class="modal-card-title flex items-center justify-between text-xl font-weight-normal">
            <span style="width: 24px; height: 48px" />
            <span class="text-2xl flex-1 text-center font-bold capitalize">{{ header }}</span>
            <a
              href="#"
              class="text-black self-start"
              @click.prevent="closeModal"
            >
              <span class="icon">
                <FontAwesomeIcon
                  :icon="['fas', 'times']"
                />
              </span>
            </a>
          </p>
        </slot>
      </header>
      <section class="modal-card-body p-0 overflow-auto">
        <slot>
          <div class="modal-card-body bg-white p-4 md:p-6 text-sm sm:text-base md:text-lg">
            <p class="mb-4">
              {{ content }}
            </p>
          </div>
        </slot>
      </section>
      <div class="card-footer">
        <div class="card-footer-item flex-col">
          <slot name="footer">
            {{ footer }}
          </slot>
        </div>
      </div>
    </div>
  </div>
</template>

<script>
import Blur from '@/mixins/Blur'

export default {
  name: 'AccordionModal',
  mixins: [Blur],
  props: {
    header: {
      type: String,
      default: '',
    },
    content: {
      type: String,
      default: '',
    },
    footer: {
      type: String,
      default: '',
    },
  },
  methods: {
    closeModal () {
      this.$emit('close')
      this.$parent.$emit('close')
      this.unBlur()
    },
  },
}
</script>
