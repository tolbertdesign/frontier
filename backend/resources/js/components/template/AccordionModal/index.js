import AccordionModal from './AccordionModal.vue'

export default AccordionModal
