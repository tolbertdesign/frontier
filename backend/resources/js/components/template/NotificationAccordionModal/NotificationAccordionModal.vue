<template>
  <AccordionModal
    :header="lang.alerts"
    class="m-2"
  >
    <template
      v-if="notifications.length > 0"
    >
      <Accordion
        v-for="notification in notifications"
        :key="notification.id"
        :title="getTitle(notification)"
        :is-open="notifications.length === 1"
        :should-emit-viewed="true"
        :title-font-weight="fontWeight(notification)"
        animation="zoom-out"
        @viewed="updateNotification(notification)"
      >
        <template>
          <div class="p-2">
            <div class="modal-card-body bg-white p-4 md:p-6 text-sm sm:text-base md:text-lg">
              <p
                class="mb-4"
                v-html="content(notification)"
              />
            </div>
          </div>
        </template>
      </Accordion>
    </template>
    <div
      v-else
      class="py-2 text-center"
    >
      {{ lang.no_alerts }}
    </div>
  </AccordionModal>
</template>

<script>
import AccordionModal from '@/components/template/AccordionModal'
import Accordion from '@/components/element/Accordion'
import Blur from '@/mixins/Blur'

export default {
  name: 'NotificationAccordionModal',
  components: {
    AccordionModal,
    Accordion,
  },
  mixins: [Blur],
  props: {
    program: {
      type: Object,
      default: () => {},
    },
  },
  computed: {
    lang () {
      return this.$store.state.lang
    },
    notifications () {
      return this.$store.getters.sortNotificationsByProgramId(this.program.id)
    },
  },
  created () {
    if (this.$store.state.notification.notifications.length === 1) {
      this.updateNotification(this.$store.state.notification.notifications[0])
    }
  },
  methods: {
    closeModal () {
      this.$emit('close')
      this.unBlur()
    },
    updateNotification (notification) {
      if (!this.$store.getters.isRead(notification)) {
        this.$store.dispatch('readNotification', notification)
      }
    },
    fontWeight (notification) {
      if (notification.read_at) {
        return 'font-normal'
      }

      return 'font-semibold'
    },
    content (notification) {
      return notification.clean_html_message || ''
    },
    getTitle (notification) {
      const program = this.$store.state.User.programs.find(program => program.id === notification.program_id)
      return program.event_name
    },
  },
}
</script>
