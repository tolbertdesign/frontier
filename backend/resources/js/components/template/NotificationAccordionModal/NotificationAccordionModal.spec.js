import { shallowMount, createLocalVue } from '@vue/test-utils'
import NotificationAccordionModal from './NotificationAccordionModal.vue'
import AccordionModal from '@/components/template/AccordionModal'
import Accordion from '@/components/element/Accordion'
import Blur from '@/mixins/Blur'

const localVue = createLocalVue()

const id = 1
const read_at = '2020-01-01 00:00:00'
const clean_html_message = 'clean HTML message'
const type = 'type'
const program_id = 10
const notification = {
  id: id,
  program_id: program_id,
  read_at: read_at,
  clean_html_message: clean_html_message,
  type: type,
}
const notification2 = {
  id: 3,
  program_id: program_id,
  read_at: null,
  clean_html_message: clean_html_message,
  type: type,
  $storeActionCalled: null,
}
const notifications = [
  notification,
]
const eventName = 'event name'
const program = {
  id: program_id,
  event_name: eventName,
}
const $store = {
  state: {
    lang: {
      alerts: 'alerts',
      program_alert: 'program_alert',
    },
    notification: {
      TYPE_PROGRAM: type,
      notifications: [notification],
    },
    User: {
      programs: [program],
    },
  },
  getters: {
    isRead: jest.fn(() => false),
    sortNotificationsByProgramId: jest.fn(id => [notification]),
  },
  dispatch: jest.fn((name, notification) => {
    notification.$storeActionCalled = name
    return notification
  }),
  actions: {
    readNotification: jest.fn(),
  },
}

describe('NotificationAccordionModal', () => {
  it('Snapshop test for NotificationAccordionModal with notifications', () => {
    const wrapper = shallowMount(NotificationAccordionModal, {
      localVue,
      mixins: {
        Blur,
      },
      components: {
        Accordion,
        AccordionModal,
      },
      mocks: {
        $store,
      },
      propsData: {
        program: program,
      },
    })
    expect(wrapper).toMatchSnapshot()
  })

  it('Snapshop test for NotificationAccordionModal without notifications', () => {
    const wrapper = shallowMount(NotificationAccordionModal, {
      localVue,
      mixins: {
        Blur,
      },
      components: {
        Accordion,
        AccordionModal,
      },
      mocks: {
        $store,
      },
      propsData: {
        program: program,
      },
    })
    expect(wrapper).toMatchSnapshot()
  })

  it('Snapshop test for NotificationAccordionModal without event name', () => {
    const wrapper = shallowMount(NotificationAccordionModal, {
      localVue,
      mixins: {
        Blur,
      },
      components: {
        Accordion,
        AccordionModal,
      },
      mocks: {
        $store,
      },
      propsData: {
        program: program,
      },
    })
    expect(wrapper.vm.header).toBe($store.state.lang.notifications)
    expect(wrapper).toMatchSnapshot()
  })

  it('Snapshop test for NotificationAccordionModal with unread notification', () => {
    notification.read_at = null
    const wrapper = shallowMount(NotificationAccordionModal, {
      localVue,
      mixins: {
        Blur,
      },
      components: {
        Accordion,
        AccordionModal,
      },
      mocks: {
        $store,
      },
      propsData: {
        program: program,
      },
    })
    expect(wrapper).toMatchSnapshot()
  })

  it('calculates title correctly from the program', () => {
    const wrapper = shallowMount(NotificationAccordionModal, {
      localVue,
      mixins: {
        Blur,
      },
      components: {
        Accordion,
        AccordionModal,
      },
      mocks: {
        $store,
      },
      propsData: {
        program: program,
      },
    })
    expect(wrapper.vm.getTitle(notification)).toBe($store.state.User.programs[0].event_name)
  })

  it('calls notification action readNotification on created when only one notification in array', () => {
    notifications[0].$storeActionCalled = null
    const wrapper = shallowMount(NotificationAccordionModal, {
      localVue,
      mixins: {
        Blur,
      },
      components: {
        Accordion,
        AccordionModal,
      },
      mocks: {
        $store,
      },
      propsData: {
        program: program,
      },
    })

    wrapper.vm.updateNotification(notifications[0])
    expect(notifications[0].$storeActionCalled).toBe('readNotification')
  })

  it('calls notification action readNotification when notification is read', () => {
    notifications[0].$storeActionCalled = null
    $store.state.notification.notifications.push(notification2)
    const wrapper = shallowMount(NotificationAccordionModal, {
      localVue,
      mixins: {
        Blur,
      },
      components: {
        Accordion,
        AccordionModal,
      },
      mocks: {
        $store,
      },
      propsData: {
        program: program,
      },
    })
    expect(notifications[0].$storeActionCalled).toBe(null)
    wrapper.vm.updateNotification(notifications[0])
    expect(notifications[0].$storeActionCalled).toBe('readNotification')
  })
})
