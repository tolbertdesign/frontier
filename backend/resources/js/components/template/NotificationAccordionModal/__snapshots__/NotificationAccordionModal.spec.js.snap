// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NotificationAccordionModal Snapshop test for NotificationAccordionModal with notifications 1`] = `
<accordionmodal-stub header="alerts" content="" footer="" class="m-2">
  <accordion-stub title="event name" content="" isopen="true" shouldemitviewed="true" titlefontweight="font-normal" animation="zoom-out">
    <div class="p-2">
      <div class="modal-card-body bg-white p-4 md:p-6 text-sm sm:text-base md:text-lg">
        <p class="mb-4">clean HTML message</p>
      </div>
    </div>
  </accordion-stub>
</accordionmodal-stub>
`;

exports[`NotificationAccordionModal Snapshop test for NotificationAccordionModal with unread notification 1`] = `
<accordionmodal-stub header="alerts" content="" footer="" class="m-2">
  <accordion-stub title="event name" content="" isopen="true" shouldemitviewed="true" titlefontweight="font-semibold" animation="zoom-out">
    <div class="p-2">
      <div class="modal-card-body bg-white p-4 md:p-6 text-sm sm:text-base md:text-lg">
        <p class="mb-4">clean HTML message</p>
      </div>
    </div>
  </accordion-stub>
</accordionmodal-stub>
`;

exports[`NotificationAccordionModal Snapshop test for NotificationAccordionModal without event name 1`] = `
<accordionmodal-stub header="alerts" content="" footer="" class="m-2">
  <accordion-stub title="event name" content="" isopen="true" shouldemitviewed="true" titlefontweight="font-normal" animation="zoom-out">
    <div class="p-2">
      <div class="modal-card-body bg-white p-4 md:p-6 text-sm sm:text-base md:text-lg">
        <p class="mb-4">clean HTML message</p>
      </div>
    </div>
  </accordion-stub>
</accordionmodal-stub>
`;

exports[`NotificationAccordionModal Snapshop test for NotificationAccordionModal without notifications 1`] = `
<accordionmodal-stub header="alerts" content="" footer="" class="m-2">
  <accordion-stub title="event name" content="" isopen="true" shouldemitviewed="true" titlefontweight="font-normal" animation="zoom-out">
    <div class="p-2">
      <div class="modal-card-body bg-white p-4 md:p-6 text-sm sm:text-base md:text-lg">
        <p class="mb-4">clean HTML message</p>
      </div>
    </div>
  </accordion-stub>
</accordionmodal-stub>
`;
